### PR TITLE
Fix SR anomaly score calculation at beginning

### PR DESCRIPTION
**Pull request by guinao**
We are excited to review your PR.

So we can do the best job, please check:

- [ ] There's a descriptive title that will make sense to other developers some time from now. 
- [ ] There's associated issues. All PR's should have issue(s) associated - unless a trivial self-evident change such as fixing a typo. You can use the format `Fixes #nnnn` in your description to cause GitHub to automatically close the issue(s) when your PR is merged.
- [ ] Your change description explains what the change does, why you chose your approach, and anything else that reviewers should know.
- [ ] You have included any necessary tests in the same PR.

#5491

### CODE REVIEW

**Comment by guinao** [comment]
@mengai is added to the review. #Closed

**Comment by guinao** [comment]
@conhua is added to the review. #Closed

**Comment by guinao** [comment]
@jinhua is added to the review. #Closed

**Comment by guinao** [comment]
@nnoradie is added to the review. #Closed

**Comment by guinao** [comment]
@harishsk is added to the review. #Closed

File: src/Microsoft.ML.TimeSeries/SrCnnEntireAnomalyDetector.cs [hunk 1]
```diff
@@ -625,7 +649,7 @@ private double PredictNext(double[] data)
                 return (data[1] + slopeSum);
             }
 
-            private void AverageFilter(double[] data, int n)
+            private void AverageFilter(double[] data, int n, bool ignoreFirst=false)
```
Comment (lisahua) [inline src/Microsoft.ML.TimeSeries/SrCnnEntireAnomalyDetector.cs:652]: where this `ignoreFirst` being used?

File: src/Microsoft.ML.TimeSeries/SrCnnEntireAnomalyDetector.cs [hunk 2]
```diff
@@ -351,6 +360,15 @@ internal sealed class SrCnnEntireModeler
             private static readonly double _deanomalyThreshold = 0.35;
             private static readonly double _boundSensitivity = 93.0;
             private static readonly double _unitForZero = 0.3;
+            private static readonly double _minimumScore = 0.0;
+            private static readonly double _maximumScore = 1.0;
+            // If the score window is smaller than this value, the anomaly score is tend to be small.
```
Comment (lisahua) [inline src/Microsoft.ML.TimeSeries/SrCnnEntireAnomalyDetector.cs:365]: still a bit confused at `score window`, is it the same as `batchSize`?

File: src/Microsoft.ML.TimeSeries/SrCnnEntireAnomalyDetector.cs [hunk 3]
```diff
@@ -309,6 +309,15 @@ public void Process()
                     _previousBatch = _previousBatch.GetRange(_batch.Count, _bLen);
                     _previousBatch.AddRange(_batch);
                     _modeler.Train(_previousBatch.ToArray(), ref _results);
+
+                    // move the values to front
+                    for (int i = 0; i < _batch.Count; ++i)
```
Comment (lisahua) [inline src/Microsoft.ML.TimeSeries/SrCnnEntireAnomalyDetector.cs:314]: This is only for the first batch right

File: src/Microsoft.ML.TimeSeries/SrCnnEntireAnomalyDetector.cs (hunk 2 above)
Comment (lisahua) [inline src/Microsoft.ML.TimeSeries/SrCnnEntireAnomalyDetector.cs:365]: just curious,` score window` is the same as `batchSize`?

File: src/Microsoft.ML.TimeSeries/SrCnnEntireAnomalyDetector.cs (hunk 1 above)
Comment (lisahua) [inline src/Microsoft.ML.TimeSeries/SrCnnEntireAnomalyDetector.cs:652]: is `ignoreFirst` ever used?

**Review by lisahua — APPROVED** [review]
Pending for Nneoma to confirm it locally

**Comment by nnoradie** [comment]
image
 nice, looks good

**Comment by nnoradie** [comment]
Looks we are finding some new anomaly points in the first two points as expected with this fix. Along with some minor boundary changes through the rest of the results. 
image

File: src/Microsoft.ML.TimeSeries/SrCnnEntireAnomalyDetector.cs (hunk 1 above)
Comment (guinao) [reply src/Microsoft.ML.TimeSeries/SrCnnEntireAnomalyDetector.cs]: Removed this parameter, it is unused actually.

---
In reply to: [527820832](https://github.com/dotnet/machinelearning/pull/5502#discussion_r527820832) [](ancestors = 527820832)

File: src/Microsoft.ML.TimeSeries/SrCnnEntireAnomalyDetector.cs (hunk 2 above)
Comment (guinao) [reply src/Microsoft.ML.TimeSeries/SrCnnEntireAnomalyDetector.cs:365]: It is not the same as batchSize, it is the `_judgementWindowSize`.

---
In reply to: [527820502](https://github.com/dotnet/machinelearning/pull/5502#discussion_r527820502) [](ancestors = 527820502)

File: src/Microsoft.ML.TimeSeries/SrCnnEntireAnomalyDetector.cs (hunk 3 above)
Comment (guinao) [reply src/Microsoft.ML.TimeSeries/SrCnnEntireAnomalyDetector.cs:314]: This code snippet is for another bug, which is the result not aligned for the last batch. It is not related with the reported case.

---
In reply to: [527820005](https://github.com/dotnet/machinelearning/pull/5502#discussion_r527820005) [](ancestors = 527820005)

**Comment by guinao** [comment]
Yes, it adds up the likelyhood that the first two points recognized as anomaly.

---
In reply to: [731465154](https://github.com/dotnet/machinelearning/pull/5502#issuecomment-731465154) [](ancestors = 731465154)

**Comment by guinao** [comment]
cool

---
In reply to: [731324217](https://github.com/dotnet/machinelearning/pull/5502#issuecomment-731324217) [](ancestors = 731324217)

**Review by harishsk — APPROVED** [review]
:shipit:

**Pull request merged**